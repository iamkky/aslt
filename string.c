#include <stdlib.h>
#include <string.h>
#include <stdarg.h>
#include <stdio.h>
#include <ctype.h>

#define CSIZE(class)    (sizeof(*(class)0))

typedef struct StList_struct *StList;
struct StList_struct {
	char **list;
	int size;
	StList jj;
};

typedef struct StBuffer_struct *StBuffer;
struct StBuffer_struct {
	char *buffer;
	int  allocated;
	int  size;
};

StList stListNew(int size)
{
StList self;

	if((self=malloc(CSIZE(StList)))==NULL) return NULL;

	self->size = size;
	if((self->list = malloc(sizeof(char *) * self->size)) == NULL){
		free(self);
		return NULL;
	}

	self->list[0] = NULL;

	return self;
}

int stListRegister(StList self, char *str)
{
char **tmp;
int c;

	if(self==NULL) return -1;

	for(c=0; c<=self->size; c++){
		if(self->list[c]==NULL) break;
		if(strcmp(self->list[c],str)==0) return c;
	}

	if(c + 1>=self->size){
		if((tmp = realloc(self->list, sizeof(char *) * self->size * 2))==NULL) return -1;
		self->size *= 2;
		self->list = tmp;
	}

	self->list[c] = strdup(str);
	self->list[c + 1] = NULL;
}

int stListToupper(StList self)
{
char	*p;
int	c;

	for(c=0; c<self->size; c++){
                if(self->list[c] == NULL) break;
		for(p=self->list[c]; *p; p++) *p = toupper(*p);
        }	
	return 0;
}

StBuffer stBufferNew(int alloc)
{
StBuffer self;

	if((self=malloc(CSIZE(StBuffer)))==NULL) return NULL;

	if(alloc <= 0) alloc = 128;

	self->size = 0;
	self->allocated = alloc;
	self->buffer = malloc(sizeof(char) * self->allocated);

	if(self->buffer == NULL){
		free(self);
		return NULL;
	}

	self->buffer[0] = 0;

	return self;
}

int stBufferAppend(StBuffer self, char *str)
{
char *tmp;
int len1, len2, s;

	if(self==NULL) return -1;

	len1 = strlen(self->buffer);
	len2 = strlen(str);

	if(len1 + len2 + 1 >= self->allocated){
		s = self->allocated;
		while(len1 + len2 + 1 > s) s *= 2;
		if((tmp = realloc(self->buffer, sizeof(char *) * s))==NULL) return -1;
		self->allocated = s;
		self->buffer = tmp;
	}

	strcpy(self->buffer + len1, str);
	self->size = len1 + len2;

	return len2;
}

int stBufferAppendf(StBuffer self, char *fmt, ...)
{
va_list va;
char *tmp;
int len1, len2, s;

	if(self==NULL) return -1;

	va_start(va, fmt);

	len1 = strlen(self->buffer);
	len2 = vsnprintf(self->buffer + len1, self->allocated - len1, fmt, va);

	if(len1 + len2 + 1 >= self->allocated){
		s = self->allocated;
		while(len1 + len2 + 1 > s) s *= 2;
		if((tmp = realloc(self->buffer, sizeof(char *) * s))==NULL) return -1;
		self->allocated = s;
		self->buffer = tmp;
	}

	len2 = vsnprintf(self->buffer + len1, self->allocated - len1, fmt, va);
	self->size = len1 + len2;

	return len2;
}

