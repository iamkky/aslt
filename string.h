#ifndef _RDPP_STRING_H_
#define _RDPP_STRING_H_

#define CSIZE(class)    (sizeof(*(class)0))

typedef struct StList_struct *StList;
struct StList_struct {
	char **list;
	int size;
	StList jj;
};

StList stListNew(int size);
int stListRegister(StList self, char *str);
int stListToupper(StList self);

typedef struct StBuffer_struct *StBuffer;
struct StBuffer_struct {
	char *buffer;
	int  allocated;
	int  size;
};

StBuffer stBufferNew(int alloc);
int stBufferAppend(StBuffer self, char *str);
int stBufferAppendf(StBuffer self, char *fmt, ...);

#endif
