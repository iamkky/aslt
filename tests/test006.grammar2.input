

rule: nonterminal COLON rside SEMICOLON {
		if(x>0) {
			printf("check \"%s\"\n", checkname);
		}
	};

rside: itemlist code rsideor {
		fprintf(stderr,"%s \n", checkname);
	};

rsideor: VBAR rside | ;

itemlist: item itemlist | ;

item: NONTERMINAL | TERMINAL {
	// Last code
	return NULL
	};

code: CODEBETWEENCURLYBRACES | ;


