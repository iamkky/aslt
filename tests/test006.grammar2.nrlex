#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>
#include "test_utils.h"

/*
rule: nonterminal COLON rside SEMICOLON ;

rside: item_list code rside_or ;

rside_or: VBAR rside | ;

item_list: item item_list | ;

item: NONTERMINAL | TERMINAL ;

code: CODEBETWEENCURLYBRACES | empty ;
*/

enum TOKENS {COLON=1000, SEMICOLON, VBAR, TERMINAL, NONTERMINAL, CODEBETWEENCURLYBRACES};
char *tokens[]= {"COLON", "SEMICOLON", "VBAR", "TERMINAL", "NONTERMINAL", "CODEBETWEENCURLYBRACES"};

int basic_lex(char *buffer, int *cursor, char* *value, void *extra);

int lcount;

int main(int argc, char **argv)
{
char *value;
char *buffer;
int  size;
int  cursor;
int  token;

	buffer = readToBuffer(0, 128, 1, &size);
	buffer[size]=0;
	
	cursor = 0;
	lcount = 0;
	
	do{
		token=basic_lex(buffer,&cursor,&value,NULL);
		fprintf(stdout, "Token: %d %s\n", token, token>0 ? tokens[token-1000] : "Invalid");
		switch(token){
		case TERMINAL:
		case NONTERMINAL:
		case CODEBETWEENCURLYBRACES:
			fprintf(stdout,"Value: %s\n", value);
		}
	}while(token>=0);

	if(*(buffer+cursor)){
		fprintf(stdout,"Error line %d: %20.20s\n", lcount, buffer+cursor);
	};
}

// Lex Helper Function

static char *getCode(char *input, int *cursor)
{
char *tmp, *p, last;
int  count, level, string, character, escaped;

	if(*(p=input++)!='{') return NULL;

	count = level = string = character = escaped = 0;

	while((last = *input++)){
		count++;
		if(escaped){
			escaped = 0;
		}else if(character==0 && last=='"'){
			if(string == 0) string = 1; else string = 0;
		}else if(string==1 && last=='\\'){
			escaped = 1;
		}else if(string==0 && last=='\''){
			if(character == 0) character = 1; else character = 0;
		}else if(character==1 && last=='\\'){
			escaped = 1;
		}else if(!string && !character && last=='{'){
			level++;
		}else if(!string && !character && last=='}'){
			if(level--==0) break;
		}
	}

	if((tmp = malloc(count))==NULL) return NULL;
	strncpy(tmp, p+1, count-1);

	(*cursor) += count;
	return tmp;
}

%%basic_lex char*

%define a a-z
%define A A-Z

:		: return COLON;
;		: return SEMICOLON;
|		: return VBAR;
\a\a*		: *value = nrDupToken(input, size); return NONTERMINAL;
\A\A*		: *value = nrDupToken(input, size); return TERMINAL;
{		: *value = getCode(input, cursor); return CODEBETWEENCURLYBRACES;
\s		: 
\n		: lcount++;
\t		: 
\r		: 
