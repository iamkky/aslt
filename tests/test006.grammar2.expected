Token: 1004 NONTERMINAL
Value: rule
Token: 1000 COLON
Token: 1004 NONTERMINAL
Value: nonterminal
Token: 1003 TERMINAL
Value: COLON
Token: 1004 NONTERMINAL
Value: rside
Token: 1003 TERMINAL
Value: SEMICOLON
Token: 1005 CODEBETWEENCURLYBRACES
Value: 
		if(x>0) {
			printf("check \"%s\"\n", checkname);
		}
	
Token: 1001 SEMICOLON
Token: 1004 NONTERMINAL
Value: rside
Token: 1000 COLON
Token: 1004 NONTERMINAL
Value: itemlist
Token: 1004 NONTERMINAL
Value: code
Token: 1004 NONTERMINAL
Value: rsideor
Token: 1005 CODEBETWEENCURLYBRACES
Value: 
		fprintf(stderr,"%s \n", checkname);
	
Token: 1001 SEMICOLON
Token: 1004 NONTERMINAL
Value: rsideor
Token: 1000 COLON
Token: 1003 TERMINAL
Value: VBAR
Token: 1004 NONTERMINAL
Value: rside
Token: 1002 VBAR
Token: 1001 SEMICOLON
Token: 1004 NONTERMINAL
Value: itemlist
Token: 1000 COLON
Token: 1004 NONTERMINAL
Value: item
Token: 1004 NONTERMINAL
Value: itemlist
Token: 1002 VBAR
Token: 1001 SEMICOLON
Token: 1004 NONTERMINAL
Value: item
Token: 1000 COLON
Token: 1003 TERMINAL
Value: NONTERMINAL
Token: 1002 VBAR
Token: 1003 TERMINAL
Value: TERMINAL
Token: 1005 CODEBETWEENCURLYBRACES
Value: 
	// Last code
	return NULL
	
Token: 1001 SEMICOLON
Token: 1004 NONTERMINAL
Value: code
Token: 1000 COLON
Token: 1003 TERMINAL
Value: CODEBETWEENCURLYBRACES
Token: 1002 VBAR
Token: 1001 SEMICOLON
Token: -1 Invalid
